enterprise/wsproxy/keycache.go

```diff
@@ -0,0 +1,225 @@
+package wsproxy
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/cryptokeys"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/quartz"
+)
+
+const (
+	// latestSequence is a special sequence number that represents the latest key.
+	latestSequence = -1
+	// refreshInterval is the interval at which the key cache will refresh.
+	refreshInterval = time.Minute * 10
+)
+
+type Fetcher interface {
+	Fetch(ctx context.Context) ([]codersdk.CryptoKey, error)
+}
+
+type CryptoKeyCache struct {
+	Clock         quartz.Clock
+	refreshCtx    context.Context
+	refreshCancel context.CancelFunc
+	fetcher       Fetcher
+	logger        slog.Logger
+
+	mu        sync.Mutex
+	keys      map[int32]codersdk.CryptoKey
+	latest    codersdk.CryptoKey
+	lastFetch time.Time
+	refresher *quartz.Timer
+	fetching  bool
+	closed    bool
+	cond      *sync.Cond
+}
+
+func NewCryptoKeyCache(ctx context.Context, log slog.Logger, client Fetcher, opts ...func(*CryptoKeyCache)) (*CryptoKeyCache, error) {
+	cache := &CryptoKeyCache{
+		Clock:   quartz.NewReal(),
+		logger:  log,
+		fetcher: client,
+	}
+
+	for _, opt := range opts {
+		opt(cache)
+	}
+
+	cache.cond = sync.NewCond(&cache.mu)
+	cache.refreshCtx, cache.refreshCancel = context.WithCancel(ctx)
+	cache.refresher = cache.Clock.AfterFunc(refreshInterval, cache.refresh)
+
+	m, latest, err := cache.cryptoKeys(ctx)
+	if err != nil {
+		cache.refreshCancel()
+		return nil, xerrors.Errorf("initial fetch: %w", err)
+	}
+	cache.keys, cache.latest = m, latest
+
+	return cache, nil
+}
+
+func (k *CryptoKeyCache) Signing(ctx context.Context) (codersdk.CryptoKey, error) {
+	return k.cryptoKey(ctx, latestSequence)
+}
+
+func (k *CryptoKeyCache) Verifying(ctx context.Context, sequence int32) (codersdk.CryptoKey, error) {
+	return k.cryptoKey(ctx, sequence)
+}
+
+func (k *CryptoKeyCache) cryptoKey(ctx context.Context, sequence int32) (codersdk.CryptoKey, error) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	if k.closed {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	var key codersdk.CryptoKey
+	var ok bool
+	for key, ok = k.key(sequence); !ok && k.fetching && !k.closed; {
+		k.cond.Wait()
+	}
+
+	if k.closed {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	if ok {
+		return checkKey(key, sequence, k.Clock.Now())
+	}
+
+	k.fetching = true
+	k.mu.Unlock()
+
+	keys, latest, err := k.cryptoKeys(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("get keys: %w", err)
+	}
+
+	k.mu.Lock()
+	k.lastFetch = k.Clock.Now()
+	k.refresher.Reset(refreshInterval)
+	k.keys, k.latest = keys, latest
+	k.fetching = false
+	k.cond.Broadcast()
+
+	key, ok = k.key(sequence)
+	if !ok {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return checkKey(key, sequence, k.Clock.Now())
+}
+
+func (k *CryptoKeyCache) key(sequence int32) (codersdk.CryptoKey, bool) {
+	if sequence == latestSequence {
```

Comment:
if you're going to use a special int32 to represent latest, you might as well go whole hog and just insert the crypto key into the map under that special value, and drop `k.latest`